cpmpy/expressions/globalconstraints.py

```diff
@@ -280,21 +280,26 @@ def decompose(self):
         succ = cpm_array(self.args)
         n = len(succ)
         order = intvar(0,n-1, shape=n, name='order')
+        defining = []
         constraining = []
         constraining += [AllDifferent(succ)] # different successors
-        constraining += [AllDifferent(order)] # different orders
-        constraining += [order[n-1] == 0] # symmetry breaking, last one is '0'
-        a = boolvar(name='a') # true iff the values in succ are within bounds.
+        constraining += [AllDifferent(order)]  # different orders
+        constraining += [order[n-1] == 0]  # symmetry breaking, last one is '0'
+        a = boolvar(name='a')  # true iff the values in succ are within bounds.
         # Otherwise they lose meaning (it means nothing to come after the 9th element of a list with only 7 elements)
-
-        defining = [a == ((Minimum(succ) >= 0) & (Maximum(succ) < n))]
-        for i in range(n):
-            defining += [(~a).implies(order[i] == 0)] # assign arbitrary value, no ordering exists if ~a
-            if i == 0:
-                defining += [a.implies(order[0] == succ[0])]
-            else:
-                defining += [a.implies(
-                    order[i] == succ[order[i - 1]])]  # first one is successor of '0', ith one is successor of i-1
+        # if all bounds are tight, we can just set a = true
+        lb, ub = get_bounds(succ)
+        if min(lb) >= 0 and max(ub) < n:
```

Comment:
Meh, I would be more explicit here on what is going on...
I.e., smt like:

```python3

lbs, ubs = get_bounds(self.args)
if min(lbs) > 0 or max(ubs) < n-1: # can never be satisfied
    return [BoolVar(False)], []
elif min(lbs) >= 0 and max(ubs) < n:
    is_defined=BoolVal(True) # bounds ensure it is always defined
else: # bounds may exceed length of array or be negative
    is_defined = boolvar()
...
```
And also an elaborate docstring on why we do this would be handy.
First attemp:

```python3
"""
The order-variables denote the ordering of the nodes in the cycle. I.e., `order[i] == succ[order[i-1]]`. 
However, the next node must only be defined when the Circuit is asserted to be True.
Hence, `~Circuit(succ)` can be satisfied if there is subcircuit, 
or if the values of the variables do not define a successor function. 
That is, for some i, `succ[i] < 0 or succ[i] >= len(succ)`.
In the decomposition, we handle this explicitly by introducing an `is_defined` indicator variable.
"""
```